Source/Core/DolphinQt2/Config/GeckoCodeWidget.cpp

```diff
@@ -0,0 +1,175 @@
+
+  UpdateList();
+}
+
+void GeckoCodeWidget::CreateWidgets()
+{
+  auto* layout = new QVBoxLayout;
+  m_warning = new CheatWarningWidget(m_game_id);
+  m_code_list = new QListWidget;
+  m_code_description = new QTextEdit;
+  m_code_view = new QListWidget;
+  m_download_codes = new QPushButton(tr("Download Codes (WiiRD Database)"));
+
+  m_code_description->setReadOnly(true);
+
+  m_download_codes->setEnabled(!m_game_id.empty());
+
+  layout->addWidget(m_warning);
+  layout->addWidget(m_code_list, 3);
+  layout->addWidget(m_code_description, 1);
+  layout->addWidget(m_code_view, 1);
+  layout->addWidget(m_download_codes);
+
+  setLayout(layout);
+}
+
+void GeckoCodeWidget::ConnectWidgets()
+{
+  connect(m_code_list, &QListWidget::itemSelectionChanged, this,
+          &GeckoCodeWidget::OnSelectionChanged);
+  connect(m_code_list, &QListWidget::itemChanged, this, &GeckoCodeWidget::OnItemChanged);
+
+  connect(m_download_codes, &QPushButton::pressed, this, &GeckoCodeWidget::DownloadCodes);
+
+  connect(m_warning, &CheatWarningWidget::OpenCheatEnableSettings, this,
+          &GeckoCodeWidget::OpenGeneralSettings);
+
+  connect(this, &GeckoCodeWidget::EmulationStarted, m_warning,
+          &CheatWarningWidget::EmulationStarted);
+  connect(this, &GeckoCodeWidget::EmulationStopped, m_warning,
+          &CheatWarningWidget::EmulationStopped);
+}
+
+void GeckoCodeWidget::OnSelectionChanged()
+{
+  auto items = m_code_list->selectedItems();
+
+  if (items.empty())
+    return;
+
+  auto* selected = items[0];
+
+  auto& code = m_gecko_codes[m_code_list->row(selected)];
+
+  QString notes;
+
+  for (const std::string& line : code.notes)
+    notes.append(QString::fromStdString(line));
+
+  m_code_description->setHtml(
+      tr("<pre><b>Creator:</b> %1\n<b>Description:</b>\n\n%2</pre>")
+          .arg(QString::fromStdString(code.creator), notes.isEmpty() ? tr("N/A") : notes));
+
+  m_code_view->clear();
+
+  QFont monospace(QStringLiteral("Monospace"));
+  monospace.setStyleHint(QFont::TypeWriter);
+
+  m_code_view->setFont(monospace);
+
+  for (const auto& c : code.codes)
+    m_code_view->addItem(QStringLiteral("%1  %2")
+                             .arg(c.address, 8, 16, QStringLiteral("0").at(0))
+                             .arg(c.data, 8, 16, QStringLiteral("0").at(0)));
+}
+
+void GeckoCodeWidget::OnItemChanged(QListWidgetItem* item)
+{
+  m_gecko_codes[m_code_list->row(item)].enabled = (item->checkState() == Qt::Checked);
+
+  m_game_ini_local = SConfig::GetInstance().LoadLocalGameIni(m_game_id, m_game_revision);
+
+  Gecko::SaveCodes(m_game_ini_local, m_gecko_codes);
+
+  auto rev_filename = m_game_id + std::to_string(m_game_revision) + ".ini";
+
+  auto filename = File::Exists(File::GetUserPath(D_GAMESETTINGS_IDX) + rev_filename) ?
+                      rev_filename :
+                      m_game_id + ".ini";
+
+  m_game_ini_local.Save(File::GetUserPath(D_GAMESETTINGS_IDX) + filename);
```

Comment:
Hmm, this logic seems like it might exist somewhere else already, and if not it probably should? (like `Config::GetLocalGameIniFilename` or something).

Also I gotta ask... why so many newlines? It's like reading an essay where each sentence is a new paragraph 😜

Reply:
It currently doesn't, sadly.

About the newlines: I try to start a new block for variable definitions and there are a lot here so it ends up like this.

Reply:
Okay, looking at this further, this seems like new behavior compared to DolphinWX. Why?

(Also, the filename-with-revision isn't correct: it should be `${gameid}r${revision}`.)

Reply:
It does it: https://github.com/dolphin-emu/dolphin/blob/100b0a52d35980c151bf205f88d8f4ae52bd0d76/Source/Core/DolphinWX/ISOProperties/ISOProperties.cpp#L657

Just with worse naming: 
https://github.com/dolphin-emu/dolphin/blob/100b0a52d35980c151bf205f88d8f4ae52bd0d76/Source/Core/DolphinWX/ISOProperties/ISOProperties.cpp#L206

Reply:
I mean that saving it to the gameid-with-revision file is new behavior: DolphinWX unconditionally saves it to the gameid-only file. Is this an intended change?

Reply:
Yes it is.